matrix-bot/src/Network/Matrix/Bot/Router/Internal.hs

```diff
@@ -0,0 +1,123 @@
+import           Control.Monad.Trans.Class         ( MonadTrans, lift )
+import           Control.Monad.Trans.State         ( StateT, get, put )
+import           Control.Monad.Trans.Writer.Lazy   ( WriterT
+                                                   , runWriterT
+                                                   , tell
+                                                   )
+import           Network.Matrix.Client             ( SyncResult(SyncResult, srNextBatch) )
+
+import           Network.Matrix.Bot.Async.Internal
+import           Network.Matrix.Bot.Event.Internal
+import           Network.Matrix.Bot.State
+
+class (MonadSyncGroupManager m) => MonadEventRouter m where
+    routeSyncEvent :: (forall n.
+                       (MonadMatrixBotBase n, MonadResyncableMatrixBot n)
+                       => a
+                       -> n ())
+                   -> a
+                   -> m ()
+    default routeSyncEvent
+        :: (m ~ m1 m2, MonadTrans m1, MonadEventRouter m2)
+        => (forall n.
+            (MonadMatrixBotBase n, MonadResyncableMatrixBot n)
+            => a
+            -> n ())
+        -> a
+        -> m ()
+    routeSyncEvent f = lift . routeSyncEvent f
+
+    routeSyncGroupEvent :: SyncGroup a -> a -> m ()
+    default routeSyncGroupEvent
+        :: (m ~ m1 m2, MonadTrans m1, MonadEventRouter m2)
+        => SyncGroup a
+        -> a
+        -> m ()
+    routeSyncGroupEvent g = lift . routeSyncGroupEvent g
+
+routeAsyncEvent :: (MonadEventRouter m)
+                => (forall n.
+                    (MonadMatrixBotBase n, MonadResyncableMatrixBot n)
+                    => a
+                    -> n ())
+                -> a
+                -> m ()
+routeAsyncEvent handle e = do
+    syncGroup <- newSyncGroup $ asyncHandler handle
+    routeSyncGroupEvent syncGroup e
+
+-- | An event router with state @s@ and acces to an environment @r@
+-- that is executed on top of the monad @m@.
+data BotEventRouter m = forall s. BotEventRouter
+    { initializeRouterState :: m s
+    , berDoRoute :: s -> BotEvent -> RouterT m s
+    }
+
+newtype RouterT m a = RouterT { unRouterM :: WriterT [SyncGroupCall m] m a }
+    deriving ( Functor, Applicative, Monad, MonadIO, MonadThrow, MonadCatch
+             , MonadMask )
+
+instance MonadTrans RouterT where
+    lift = RouterT . lift
+
+instance MonadMatrixBot m => MonadMatrixBot (RouterT m)
+
+instance (MonadResyncableMatrixBot m)
+    => MonadResyncableMatrixBot (RouterT m) where
+    withSyncStartedAt syncToken =
+        RouterT . withSyncStartedAt syncToken . unRouterM
+
+instance (MonadSyncGroupManager m) => MonadSyncGroupManager (RouterT m)
+
+instance ( MonadMatrixBotBase m
+         , MonadResyncableMatrixBot m
+         , MonadSyncGroupManager m
+         ) => MonadEventRouter (RouterT m) where
+    routeSyncEvent f x = RouterT $ tell [ syncCall f x ]
+
+    routeSyncGroupEvent group x =
+        RouterT $ tell [ asyncGroupCall group x Nothing ]
```

Comment:
nit: I find it a bit confusing that `SyncGroup` are actually async. Moreover there is already a `Sync` module to handle the matrix sync api. I know the whole point was to implement your concept of "sync group", but I now wonder if the router api should actually avoid the name Sync in it's definitions, for example:

```haskell
class MonadEventRouter m where
  routeEvent :: EventHandler a -> a -> m ()
  routeGroupEvent :: GroupEventHandler a -> a -> m ()
```

Reply:
Just to clarify: I'm still planning on doing this, but I wanted to do it separately from the last batch of changes and didn't have enough time to finish both changes.

Reply:
Ok, `SyncGroup`s are now called `EventGroup`s. The methods of `MonadEventRouter` are now `routeSyncEvent` and `routeGroupEvent`.

Reply:
That's great thank you! I don't have much time to review this right now, but feel free to merge this PR if that is easier for you, otherwise it's ok to keep it open if you don't mind adding more to the draft.